### PR TITLE
[Clang] Improve error for `-fsanitize=function/kcfi -mexecute-only` incompatibility

### LINKED ISSUE #117974: [Clang] Confusing diagnostic when using `-fsanitize=function -mexecute-only` flags together on ARM targets
When passing `-fsanitize=function -mexecute-only` to Clang on an ARM target, it prints the following error:
```
error: invalid argument '-fsanitize=function' not allowed with 'arm-unknown-linux-gnueabi'
```
The error happens because `-fsanitize=function` is incompatible with execute-only memory, but prints the target triple as the reason. I would expect the error message to be the following in this case:
```
error: invalid argument '-fsanitize=function' not allowed with '-mexecute-only'
```

The diagnostic is created here:
https://github.com/llvm/llvm-project/blob/69d66fafec968b17f84e30bc09faffb3dc8e5f15/clang/lib/Driver/SanitizerArgs.cpp#L416-L423

I think a good solution would be to print `-mexecute-only` (or `-mpure-code`) as the reason for the incompatibility if it was passed to the compiler driver, and print the target triple otherwise.

### PR DESCRIPTION
**Pull request by Il-Capitano**
The current error message when using the `-fsanitize=function -mexecute-only` flags together points to the target triple as the reason that `-fsanitize=function` is not allowed to be used, even when the function sanitizer is otherwise supported on the target when not using `-mexecute-only`.

The error message is improved to give `-mexecute-only` as the reason for disallowing `-fsanitize=function` if it was passed to the driver.

Fixes https://github.com/llvm/llvm-project/issues/117974

### CODE REVIEW

File: clang/lib/Driver/SanitizerArgs.cpp [hunk 1]
```diff
@@ -418,8 +418,14 @@ SanitizerArgs::SanitizerArgs(const ToolChain &TC,
...
-            D.Diag(diag::err_drv_argument_not_allowed_with)
-                << Desc << Triple.str();
+            llvm::opt::Arg *A = Args.getLastArgNoClaim(
+                options::OPT_mexecute_only, options::OPT_mno_execute_only);
+            if (A && A->getOption().matches(options::OPT_mexecute_only))
+              D.Diag(diag::err_drv_argument_not_allowed_with)
+                  << Desc << A->getAsString(Args);
+            else
+              D.Diag(diag::err_drv_argument_not_allowed_with)
+                  << Desc << Triple.str();
```
Comment (Il-Capitano) [inline clang/lib/Driver/SanitizerArgs.cpp:428]: Maybe `diag::err_drv_unsupported_opt_for_target` would be a better diagnostic in this case? That would produce
```
error: unsupported option '-fsanitize=function' for target '<triple>'
```

**Comment by Il-Capitano** [comment]
@MaskRay Can I ask you to review/add appropriate reviewers for this patch?

**Review by MaskRay — APPROVED** [review]
Yes, err_drv_unsupported_opt_for_target will be better for the incompatible triple error

**Comment by Il-Capitano** [comment]
Thanks for reviewing!

I updated the code to use `err_drv_unsupported_opt_for_target`, and added test cases for the `-mpure-code` alias of `-mexecute-only` also.

**Pull request merged**